### DISCUSSION TITLE
Is there a way to get the CTRLR Release version from a LUA method?

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#646] [Q&A] [answered]
Hi,
I came across the well-known problem with sendMidiMessageNow() being different between newer versions and the ones prior 5.4.

Is there a way to state something like : 
```

--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
sendParamValueSysex = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)


modIndex = mod:getProperty("vstIndex")
console("modIndex: "..modIndex)

If CtrlrVersion is >= 5.6.23 then -- For newer CTRLR versions

paramValueSysexMsg = CtrlrMidiMessage(MemoryBlock({0xF0, 0x47, 0x7E, 0x40, tonumber(modIndex), value, 0xF7}))
panel:sendMidiMessageNow(paramValueSysexMsg)

else -- For CTRLR legacy versions

paramValueSysexMsg = CtrlrMidiMessage({0xF0, 0x47, 0x7E, 0x40, tonumber(modIndex), value, 0xF7}) -- old syntax no Memoryblock
panel:sendMidiMessageNow(paramValueSysexMsg)

end -- Version Filter

end


```

It would help giving both syntaxes for sending messages without getting the crashes.

Thanks!

Damien

**Comment by Ted-juh** [comment]
Not helping much probably but isn't the only place where the CTRLR version numbers is found, is in the about section? 

So, in the about section it fetches the version number with, i think: ScopedPointer<Label> instanceVersion. And I don't think instance version is bound to Lua. 

But it also seems to be tied to ctrlrRevision (But ctrlrRevision points to the cmake setup, while instanceVersion probably takes the number from the projucer when compiling.)

Not behind my computer, so a bit hard to dig deeper into it 😕

**Reply by damiensellier** [reply]
I can also add a read only property in the globals for that in newer versions. But it will work only for 5.5.31 and above.
Also I can use the Legacy tag but it's not exactly the same.
I'll check the code tonight.

**Comment by damiensellier** [comment]
The variable for Ctrlr Version is "ctrlrRevision".
I'll add something to be able to determine which version it is from 5.6.31 for LUA methods.

**Comment by dnaldoog** [comment, marked as answer]
Something like this?

<pre>local major, minor = utils.getVersionMajor(), utils.getVersionMinor()
        major = tonumber(major)
        minor = tonumber(minor)
        if major > 4 then
            if minor > 3 then ....
</pre>

**Reply by damiensellier** [reply]
Thanks John! it's working : 

```
--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
myMethod = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

    local major, minor = utils.getVersionMajor(), utils.getVersionMinor()
    major = tonumber(major)
    minor = tonumber(minor)

    console("Build version Major is :"..major)
    console("Build version minor is :"..minor)

end
```

I get : 

```
LUA>> Build version Major is :5
LUA>> Build version minor is :6
```

So it's going to be easy to get those message sent on all Ctrlr Versions from now on.


So the orignal code would be :  

```
--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
myMethod = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

    modIndex = mod:getProperty("vstIndex")
    console("modIndex: "..modIndex)

    local major, minor = utils.getVersionMajor(), utils.getVersionMinor()
    major = tonumber(major)
    minor = tonumber(minor)

    console("Build version major is :"..major)
    console("Build version minor is :"..minor)

    if minor >= 6 then -- For newer CTRLR versions

        paramValueSysexMsg = CtrlrMidiMessage(MemoryBlock({0xF0, 0x47, 0x7E, 0x40, tonumber(modIndex), value, 0xF7}))
        panel:sendMidiMessageNow(paramValueSysexMsg)

    else  -- For CTRLR legacy versions

        paramValueSysexMsg = CtrlrMidiMessage({0xF0, 0x47, 0x7E, 0x40, tonumber(modIndex), value, 0xF7}) -- old syntax no Memoryblock
        panel:sendMidiMessageNow(paramValueSysexMsg)

    end -- Version Filter

end
```

**Reply by damiensellier** [reply]
Edit: considering there's no major version above 5 and we only filter on the minor value...

**Reply by damiensellier** [reply]
```
local major, minor, revision = utils.getVersionMajor(), utils.getVersionMinor() utils.getVersionRevision()
major = tonumber(major)
minor = tonumber(minor)
revision = tonumber(revision)
```

**Comment by dnaldoog** [comment]
Looks good!

I'm thinking the MemoryBlock addition would work anyway on 5.3 versions?

**Reply by damiensellier** [reply]
It's working on my side on Mac but not sure on PC. If it's alright on both, it's a bit useless since we can write with the latest syntax.

**Comment by dnaldoog** [comment]
Also you should be able to do this:

`mod:getVstIndex()`

**Comment by dnaldoog** [comment]
Here's a _class_ that can automate the process. i.e the user doesn't have to worry about which version they are on and adjust in every callback.

Because the class is initialised in the constructor, it is loaded across the session and never needs a cll to the class constructor again. It covers all possible types of Midi Message (that I can think of) including an empty parameter:

It can recognize `MemoryBlock():toHexString(1)` or `{}` or `MemoryBlock({})` or `string` or even `MemoryBlock()`:

----
## Constructor when panel loads:

<pre>
constructor = function(--[[ CtrlrInstance --]] type)
    mySend = Send 
    mySend:_init()
end
</pre>
----
## Class Definition
<pre>
Send = {} 
Send.__index = Send

setmetatable(
    Send,
    {
        __call = function(cls, ...)
            local self = setmetatable({}, cls)
            self:_init(...)
			return self
        end
    }
)

function Send:_init()
    self.version = self.getMinorVersion()
    self.threshold=3 
	-- usually would be 3 ie is version 5.3.* or 5.6.* ??
	-- this class will probably barf on Ctrlr versions 6.*

end

function Send:sendMessage(message)
    if type(message) == "table" then
        if self.version > self.threshold then
            panel:sendMidiMessageNow(CtrlrMidiMessage(MemoryBlock(message)))
            console("sent raw table with MB added " .. self.version)
        else
            panel:sendMidiMessageNow(CtrlrMidiMessage(message))
            console("sent raw table " .. self.version)
        end
    elseif type(message) == "string" then
        panel:sendMidiMessageNow(CtrlrMidiMessage(message))
        console("sent string possible (hexString(1) " .. self.version)
    elseif type(message) == "userdata" then
        panel:sendMidiMessageNow(CtrlrMidiMessage(message:toHexString(1)))
        console("sent raw MemoryBlock  " .. self.version)
    else
        console("illegal entry die silently  " .. self.version)
        return
    end
end

function Send:getMinorVersion()
    return utils.getVersionMinor()
end

function Send:getMajorVersion()
    return utils.getVersionMajor()
end


function Send:debug()
    console("self.version " .. self.version)
end

</pre>
----
## In a callback function that sends sysex:
<pre>
myMethod = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
if panel:getBootstrapState() then return end
    console(mod:getVstIndex() .. "")
    mySend:sendMessage({0xf0, 0, 0, value, 0xf7})
    mySend:sendMessage({0xb0, 0x07, value})
    mySend:sendMessage(MemoryBlock({0xf0, 0, 0, value, 0xf7}))
    mySend:sendMessage("F0 00 00 7f f7")
    local m= MemoryBlock("F0 00 00 63 f7")
    mySend:sendMessage(m:toHexString(1))
    mySend:sendMessage(m)
    mySend:sendMessage() -- die silently
    mySend:debug()
end
</pre>

#### EDIT: (slight refactoring) 
[Send Midi Class_1_2.zip](https://github.com/RomanKubiak/ctrlr/files/15492624/Send.Midi.Class_1_2.zip)
